src/authority.rs

```diff
@@ -1011,16 +1015,23 @@ impl<'host> TryFrom<&'host [u8]> for Host<'host> {
                 // IPv6
 
                 let ipv6 = &value[1..value.len() - 1];
+                let mut split_for_zone = ipv6.splitn(2, |c| c == &b'%');
+                let ipv6 = split_for_zone.next()
+                    .expect("Even an empty string splits into at least one component");
+                let zone = split_for_zone.next()
+                    .unwrap_or(b"");
 
-                if !check_ipv6(ipv6) {
+                if !check_ipv6(ipv6) || !check_zone(zone) {
```

Comment:
I think it'd be better if this condition was done separately from `check_ipv6` and a different error type was returned, something like `InvalidIPv6ZoneIDCharacter`